src/compiler/checker.ts

```diff
@@ -8228,6 +8358,27 @@ module ts {
                             break;
                     }
                 }
+
+                if (node.type) {
+                    if (languageVersion >= ScriptTarget.ES6 && isSyntacticallyValidGenerator(node)) {
+                        let returnType = getTypeFromTypeNode(node.type);
+                        if (returnType === voidType) {
+                            error(node.type, Diagnostics.A_generator_cannot_have_a_void_type_annotation);

```

Comment:
I think a common mistake is going to be annotating with the actual element type (instead of `IterableIterator<T>`). But when compiling this:
```typescript
function *foo(): number {
    yield 5;
}
```
I get the error "Type 'IterableIterator<any>' is not assignable to type 'number'". That seems like an odd experience. I would have expected something more along the lines of "'number' is not a proper iterator type".